gocourse03/hw/main.go

```diff
@@ -0,0 +1,327 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"os"
+    "reflect"
+    "runtime"
+	"strings"
+)
+
+func GetFunctionName(i interface{}) string {
+	fullName := runtime.FuncForPC(reflect.ValueOf(i).Pointer()).Name()
+	_, name, _ := strings.Cut(fullName, ".")
+    return name
+}
+
+type Human struct {
+	Name string
+	NameMemory bool
+	GlobalMemory bool
+	Bag bool
+}
+
+type Cave struct {
+	Entry string
+	Dark bool
+}
+
+type Backpack struct {
+	Size int
+	Matches *Matches
+	Flashlight *Flashlight
+	Knife *Knife
+}
+
+type Matches struct {
+	Item string
+	Exist bool
+}
+
+type Flashlight struct {
+	Item string
+	Exist bool
+}
+
+type Knife struct {
+	Item string
+	Exist bool
+}
+
+type Road struct {
+	Destination string
+}
+
+type Body struct {
+	IsDead bool
+	Type string
+}
+
+type Camp struct {
+	HasPeople bool
+}
+
+type Bug struct {
+	IsBig bool
+}
+
+func (h Human) Choice(choice string) {
+	fmt.Printf(choice)
+}
+
+func (b Bug) Bite() {
+	if b.IsBig {
+		fmt.Printf("Big bug bites her victim and it passes out!")
+	} else {
+		fmt.Printf("Small bug bites her victim and it just hearts!")
+	}
+} 
+
+func (h *Human) FindBag(backpack Backpack) {
+	switch h.Bag {
+	case true:
+		fmt.Printf("%s finds bag.\n", h.Name)
+		fmt.Printf("%s's bag contains:\n", h.Name)
+		if backpack.Matches.Exist == true {
+			fmt.Printf("- Matches\n") 
+		} else if backpack.Matches.Exist == false {
+			fmt.Printf("- No Matches\n") 
+		}
+		if backpack.Flashlight.Exist == true {
+			fmt.Printf("- Flashlight\n") 
+		} else if backpack.Flashlight.Exist == false {
+			fmt.Printf("- No Flashlight\n") 
+		}
+		if backpack.Knife.Exist == true {
+			fmt.Printf("- Knife\n") 
+		} else if backpack.Knife.Exist == false {
+			fmt.Printf("- No Knife\n") 
+		}
+	case false:
+		fmt.Printf("%s bag is lost.\n", h.Name)
+	}
+}
+
+func answerValidation() int {
+	var s string
+	var i int
+    fmt.Println("Enter answer number: 1 or 2")
+    for {
+		fmt.Scan(&s)
+        i, _ = strconv.Atoi(s)
+        if i == 1 || i == 2 {
+			break
+		} else {
+			fmt.Println("Enter a valid number: 1 or 2")
+		}
+	}
+	return i
+}
+
+func rememberNameChoice(remember bool, men Human) {
+	men.NameMemory = remember
+	if men.NameMemory {
+		fmt.Printf("%s remembers his name.\n", men.Name)
+	} else {
+		fmt.Printf("%s doesn't remember his name.\n", men.Name)
+	}
+}
+	
+func findBagChoice(bagFind bool, men Human, bag Backpack) {
+	men.Bag = bagFind
+	if men.Bag {
+		bag.Matches.Exist = true
```

Comment:
these changes won't be visible outside of the function since the function is only working with the copy of the structure. you need to pass a pointer for it to work properly